File: v1/LLD/Design-Patterns/Composite/code/Composite.java
// Composite Pattern — file system, uniform size() across files and folders.
// Compile: javac Composite.java
// Run:     java Composite

import java.util.ArrayList;
import java.util.List;

interface FileSystemComponent {
    int getSize();
    void print(String indent);
}

class FileLeaf implements FileSystemComponent {
    private String name;
    private int sizeKB;

    FileLeaf(String name, int sizeKB) {
        this.name = name;
        this.sizeKB = sizeKB;
    }

    public int getSize() {
        return sizeKB;
    }

    public void print(String indent) {
        System.out.println(indent + "- " + name + " (" + sizeKB + "KB)");
    }
}

class Folder implements FileSystemComponent {
    private String name;
    private List<FileSystemComponent> children = new ArrayList<>();

    Folder(String name) {
        this.name = name;
    }

    void add(FileSystemComponent component) {
        children.add(component);
    }

    public int getSize() {
        int total = 0;
        for (FileSystemComponent child : children) {
            total += child.getSize();
        }
        return total;
    }

    public void print(String indent) {
        System.out.println(indent + "+ " + name + "/ (" + getSize() + "KB total)");
        for (FileSystemComponent child : children) {
            child.print(indent + "  ");
        }
    }
}

public class Composite {
    public static void main(String[] args) {
        Folder root = new Folder("project");
        Folder src = new Folder("src");
        src.add(new FileLeaf("main.java", 12));
        src.add(new FileLeaf("utils.java", 5));

        Folder docs = new Folder("docs");
        docs.add(new FileLeaf("readme.md", 3));

        root.add(src);
        root.add(docs);
        root.add(new FileLeaf("build.gradle", 2));

        root.print("");
        System.out.println("Total size: " + root.getSize() + "KB");
    }
}

File: v1/LLD/Design-Patterns/Composite/code/composite.py
"""
Composite Pattern — file system, uniform size() across files and folders.
Run: python composite.py
"""

from abc import ABC, abstractmethod


class FileSystemComponent(ABC):
    @abstractmethod
    def get_size(self):
        ...

    @abstractmethod
    def print_tree(self, indent=""):
        ...


class FileLeaf(FileSystemComponent):
    def __init__(self, name, size_kb):
        self.name = name
        self.size_kb = size_kb

    def get_size(self):
        return self.size_kb

    def print_tree(self, indent=""):
        print(f"{indent}- {self.name} ({self.size_kb}KB)")


class Folder(FileSystemComponent):
    def __init__(self, name):
        self.name = name
        self.children = []

    def add(self, component: FileSystemComponent):
        self.children.append(component)

    def get_size(self):
        return sum(child.get_size() for child in self.children)

    def print_tree(self, indent=""):
        print(f"{indent}+ {self.name}/ ({self.get_size()}KB total)")
        for child in self.children:
            child.print_tree(indent + "  ")


if __name__ == "__main__":
    root = Folder("project")
    src = Folder("src")
    src.add(FileLeaf("main.py", 12))
    src.add(FileLeaf("utils.py", 5))

    docs = Folder("docs")
    docs.add(FileLeaf("readme.md", 3))

    root.add(src)
    root.add(docs)
    root.add(FileLeaf("setup.cfg", 2))

    root.print_tree()
    print(f"Total size: {root.get_size()}KB")
